protocol/test-infra.md

```diff
@@ -0,0 +1,358 @@
+- [optimism kurtosis](https://github.com/ethpandaops/optimism-package/) to spin up clients in a network
+
+Challenges:
+- The Go tests generally don't export to alternative implementations like Rust
+- Kurtosis is an "infra" solution to what is also a "software" problem:
+  after spinning up the network, we still need to run tests against it.
+
+This might get more approachable with more separation between node setup and testing,
+as proposed in the feature-growth section.
+Ideally we then run common tests against different combinations of each of the client implementations.
+
+### Growth of activity
+
+With more vertical scaling comes more concern around performance and stability.
+We need more benchmarks and uptime checks to understand where we can improve.
+
+Today:
+- [op-ufm (user facing monitoring)](https://github.com/ethereum-optimism/infra/tree/main/op-ufm)
+- [`replayor`](https://github.com/danyalprout/replayor)
+- Internal infra dashboards / alerts
+- One-off chain activity analysis
+
+Challenges
+- `op-ufm` should run against more chains, and be more visible to engineers.
+  E.g. if transaction inclusion speed is not good, we need to do something about it.
+- `replayor` needs automation, ideally on top of a shadow-fork of a real network,
+  such that we can analyze performance without spending real ETH or harming a production network in any way.
+- We need to monitor node performance better.
+  `pprof` / Go-resource dashboards should improve and be more center to our work.
+
+Automating flight-recording in `op-service` would be great.
+See https://go.dev/blog/execution-traces-2024 for information about flight-recording.
+E.g. on RPC call or on particular pre-programmed conditions,
+a buffer of performance data from the last N seconds can be dumped and uploaded to some service.
+When slow blocks occur, or engineers are interested in a performance snapshot of a real network, this could be great.
+Also see [op-geth flight-recording design-doc (internal)](https://github.com/ethereum-optimism/design-docs-private/blob/main/op-geth-flight-recording.md).
+
+There also exists https://pyroscope.io/ for nice continuous monitoring,
+although it is built on top of older Golang profiling hooks, and seems to provide less fine-grained information.
+
+### Growth as platform
+
+The more the stack becomes a "platform" that others fork and build new things on top on,
+the more thought we should give to the public interface and extensibility of our testing.
+
+Today: no testing platform.
+
+Challenges:
+External OP Stacks forks end up having to fork the testing infra as well, and wire in their customizations.
+A lot of tests may work by default, but some features may not.
+E.g. 4844 blob tests may need to be disabled for an alt-DA network test suite.
+
+There may be ways we can better categorize tests, make test setup more composable,
+and improve parametrization of contract versioning and chain configurations.
```

Comment:
Am I reading correctly that you're proposing we make sure hard-dependencies on some stack (ours in this case) don't creep into the test logic?
If so I'm wondering if we can enumerate the objects that potentially create such a hard-dependency, and detect their usage-patterns in the tests to ensure they're correct

Reply:
Yes, ideally we don't opinionate our own test setup / framework so much that it becomes too hard for forks of the OP-Stack to build on top of the testing.

Alt-DA variants of the OP-Stack and Alt-proof-system versions are probably the most common things those forks want to be able to test, without refactoring the core test system. A more composable test setup can go a long way here I think.

This might not be an immediate priority, but we should gather some feedback on how different OP Stack forks approach testing of their modifications, and if there are things we can do to simplify that testing.